driver-sync/src/test/functional/com/mongodb/client/AbstractTransactionsTest.java

```diff
@@ -455,4 +486,54 @@ private boolean canRunTests() {
             return false;
         }
     }
+
+    private class TargetedFailPoint {
+        private final BsonDocument failPointDocument;
+        private final MongoDatabase adminDB;
+
+        TargetedFailPoint(final BsonDocument operation) {
+            final BsonDocument arguments = operation.getDocument("arguments", new BsonDocument());
+            final ClientSession clientSession = sessionsMap.get(arguments.getString("session").getValue());
+
+            if (clientSession.getPinnedMongosAddress() != null) {
+                MongoClientSettings.Builder builder = MongoClientSettings.builder()
+                        .applyToClusterSettings(new Block<ClusterSettings.Builder>() {
+                            @Override
+                            public void apply(final ClusterSettings.Builder builder) {
+                                builder.hosts(singletonList(clientSession.getPinnedMongosAddress()));
+                            }
+                        });
+                MongoClient mongoClient = MongoClients.create(MongoClientSettings.builder(builder.build())
+                        .applyToSocketSettings(new Block<SocketSettings.Builder>() {
+                            @Override
+                            public void apply(final SocketSettings.Builder builder) {
+                                builder.readTimeout(5, TimeUnit.SECONDS);
+                            }
+                        })
+                        .build());
+
+                adminDB = mongoClient.getDatabase("admin");
+            } else {
+                adminDB = null;
+            }
+            failPointDocument = arguments.getDocument("failPoint");
+        }
+
+        public void executeFailPoint() {
+            executeCommand(failPointDocument);
+        }
+
+        public void disableFailPoint() {
+            BsonDocument disableDoc = failPointDocument.append("mode", new BsonString("off"));
+            executeCommand(disableDoc);
+        }
+
+        private void executeCommand(final BsonDocument doc) {
+            if (adminDB != null) {
+                adminDB.runCommand(doc);
+            } else {
+                collectionHelper.runAdminCommand(doc);
```

Comment:
I'm a bit concerned that this will silently do the wrong thing in the case where the session is supposed to be pinned but it's not.  Can we do something like make this assert if `useMultipleMongoses` is true but the session isn't pinned?

Reply:
Since `adminDB` is only created if there is a pinned mongos, I will assert if `useMultipleMongoses` is false before executing the `runAdminCommand`.